src/data/representation.lean

```diff
@@ -0,0 +1,312 @@
+import ring_theory.algebra
+
+/-!
+# Representations
+
+For a monoid G and a commutative ring k, there are many equivalent ways of talking about
+k-linear G-representations.
+
+1. As a typeclass `representation k G M`, extending `distrib_mul_action G M`
+  (which we introduce here).
+2. As a `module (monoid_algebra k G) M`.
+3. As a multiplicative function `G →* (M →ₗ[k] M)`.
+4. As a multiplicative function `k × G →* (M →+ M)`
+  (although note we haven't installed a monoid structure on `M →+ M` at this point).
+5. As a k-algebra morphism `monoid_algebra (k G) →ₐ[k] (M →ₗ[k] M)`.
+
+(Note the slight subtlety here that in points 1., 3., and 5. there is a "background" k-module
+structure, while in points 2. and 4. the presented data determines the k-module structure.
+For that matter, at this point we haven't even described the equivalence
+`module k M ≃ (k →* (M →+ M))`.
+In this file we define the typeclass `representation k G M`, and provide the functions
+```
+representation.as_monoid_hom [representation k G M] : G →* (M →ₗ[k] M)
+of_monoid_hom (ρ : G →* (M →ₗ[k] M)) : representation k G M
+```
+(addressing the equivalence between 1. and 3. above)
+and the functions
+```
+extend_scalars [representation k G M] : module (monoid_algebra k G) M
+representation.of_monoid_algebra_module [module (monoid_algebra k G) M] : representation k G M
+```
+(where the second function provides the `representation k G M` relative to the `module k M`
+coming from `restrict_scalars k (monoid_algebra k G) M`).
+These provide the equivalence between 1. and 2. above.
+
+The function `monoid_algebra.lift` in `data.monoid_algebra` then provides the equivalence
+between 3. and 5.
+
+(There's obviously a quite direct connection between points 2. and 5. above,
+which hasn't been developed yet.)
+
+(We don't yet say anything about point 4. above; perhaps it's less useful.)
+
+Finally, note that really all of these equivalences should be done as equivalences of categories,
+not just functions back and forth.
+We'll get there in a later file!
+
+## Tags
+
+linear action, representation
+-/
+
+universes u v w
+
+section prio
+set_option default_priority 100 -- see Note [default priority]
+/--
+A k-linear representation of a group G.
+-/
+@[nolint dangerous_instance] -- k is an `out_param`, so this is a false positive.
+class representation
+  (k : out_param $ Type v) [comm_ring k] (G : Type u) [monoid G] (M : Type w) [add_comm_group M] [module k M]
+  extends distrib_mul_action G M : Type (max u v w) :=
+(smul_smul : ∀ (g : G) (r : k) (m : M), g • (r • m) = r • (g • m))
+end prio
+
+section
+variables {k : Type v} {G : Type u} {M : Type w}
+variables [comm_ring k] [monoid G] [add_comm_group M] [module k M]
+
+/-- A group element acts by an k-linear map in any k-linear representation. -/
+def smul.linear_map [representation k G M] (g : G) : M →ₗ[k] M :=
+{ to_fun := λ m, g • m,
+  add := λ m m', by simp [smul_add],
+  smul := λ r m, by simp [representation.smul_smul], }
```

Comment:
Setting up things like this, using `representation` as a class, means that (a) there will be trouble when we want more than one action of G on M (which can surely happen) and (b) we can't say "Let pi be a representation of G on V" without breaking the rules of the class instance game. This is a big design decision. is it the right one?